src/box/alter.cc

```diff
@@ -4417,6 +4426,220 @@ on_replace_dd_schema(struct trigger * /* trigger */, void *event)
+	gc_consumer_unregister(data->new_gc);
+	return 0;
+}
+
+static int
+on_replace_dd_gc_consumers_commit(struct trigger *trigger, void *)
+{
+	struct gc_consumers_txn_trigger_data *data =
+		(struct gc_consumers_txn_trigger_data *)trigger->data;
+	struct replica *replica = replica_by_uuid(&data->uuid);
+	struct gc_consumer *new_gc = data->new_gc;
+	/* Set new GC only if the replica is still registered. */
+	if (replica != NULL) {
+		/* Drop the gc ref on creating first non-dummy consumer. */
+		if ((replica->gc == NULL || replica->gc->is_dummy) &&
+		    new_gc != NULL && !new_gc->is_dummy)
+			gc_delay_unref();
+
+		/* Drop the old gc consumer if any. */
+		if (replica->gc != NULL)
+			gc_consumer_unregister(replica->gc);
+		replica->gc = new_gc;
+	} else {
+		if (new_gc != NULL)
+			gc_consumer_unregister(new_gc);
+	}
+	return 0;
+}
+
+/**
+ * The trigger invoked on replace in space _gc_consumers.
+ * Note that due to concurrent nature of transactions it is unsafe
+ * to modify replica->gc right here and in on_rollback, so it is
+ * modified only in on_commit triggers - it is safe because they
+ * are called in order of transactions' serialization.
+ */
+static int
+on_replace_dd_gc_consumers(struct trigger * /* trigger */, void *event)
+{
+	struct txn *txn = (struct txn *)event;
+	RegionGuard region_guard(&fiber()->gc);
+	struct txn_stmt *stmt = txn_current_stmt(txn);
+	struct tuple *old_tuple = stmt->old_tuple;
+	struct tuple *new_tuple = stmt->new_tuple;
+	struct gc_consumer_def *old_def = NULL;
+	struct gc_consumer_def *new_def = NULL;
+	struct tt_uuid *replica_uuid = NULL;
+	if (old_tuple != NULL) {
+		old_def =
+			gc_consumer_def_new_from_tuple(old_tuple, &fiber()->gc);
+		if (old_def == NULL)
+			return -1;
+		replica_uuid = &old_def->uuid;
+	}
+	if (new_tuple != NULL) {
+		new_def =
+			gc_consumer_def_new_from_tuple(new_tuple, &fiber()->gc);
+		if (new_def == NULL)
+			return -1;
+		replica_uuid = &new_def->uuid;
+	}
+	assert(old_def != NULL || new_def != NULL);
+
+	/* Just making sure that both tuples have the same uuid. */
+	assert(old_def == NULL || new_def == NULL ||
+	       tt_uuid_is_equal(&old_def->uuid, &new_def->uuid));
+
+	/*
+	 * We cannot rely on the fact that the replica is still registered
+	 * in-memory because it can be dropped in the same transaction, and
+	 * replica_hash will be updated only on commit, so read row right
+	 * from the _cluster. It's important to lookup by uuid, not id,
+	 * to correctly handle the case when uuid of replica is updated
+	 * and id is the same.
+	 */
+	char key[UUID_STR_LEN + 10];
+	char *key_end = key;
+	key_end = mp_encode_array(key_end, 1);
+	key_end = mp_encode_str0(key_end, tt_uuid_str(replica_uuid));
+	assert((size_t)(key_end - key) < sizeof(key));
+	struct tuple *replica_row;
+	if (box_index_get(BOX_CLUSTER_ID, 1, key, key_end, &replica_row) != 0)
+		return -1;
+	bool replica_is_registered = replica_row != NULL;
+
+	if (!replica_is_registered && new_def != NULL) {
```

Comment:
We cannot check in-memory existence because after restart anonymous replica won't have associated `struct replica`. So this check is actually temporary, and will be removed in the next patch. If you don't like it, I can drop it right now.

Reply:
If a check isn't needed and doesn't affect any behaviour until some next patch, then lets drop it. For the sake of commits atomicity.

Reply:
Dropped the check.